command/repo.go

```diff
@@ -63,6 +80,146 @@ func repoClone(cmd *cobra.Command, args []string) error {
 	return utils.PrepareCmd(cloneCmd).Run()
 }
 
+func isURL(arg string) bool {
+	return strings.HasPrefix(arg, "http:/") || strings.HasPrefix(arg, "https:/")
+}
+
+func repoFork(cmd *cobra.Command, args []string) error {
+	ctx := contextForCommand(cmd)
+
+	forceYes, err := cmd.Flags().GetBool("yes")
+	if err != nil {
+		return err
+	}
+	forceNo, err := cmd.Flags().GetBool("no")
+	if err != nil {
+		return err
+	}
+
+	apiClient, err := apiClientForContext(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to create client: %w", err)
+	}
+
+	var toFork ghrepo.Interface
+	inParent := false // whether or not we're forking the repo we're currently "in"
+	if len(args) == 0 {
+		baseRepo, err := determineBaseRepo(cmd, ctx)
+		if err != nil {
+			return fmt.Errorf("unable to determine base repository: %w", err)
+		}
+		inParent = true
+		toFork = baseRepo
+	} else {
+		repoArg := args[0]
+
+		if isURL(repoArg) {
+			parsedURL, err := url.Parse(repoArg)
+			if err != nil {
+				return fmt.Errorf("did not understand argument: %w", err)
+			}
+
+			toFork, err = ghrepo.FromURL(parsedURL)
+			if err != nil {
+				return fmt.Errorf("did not understand argument: %w", err)
+			}
+
+		} else {
+			toFork = ghrepo.FromFullName(repoArg)
+			if toFork.RepoName() == "" || toFork.RepoOwner() == "" {
+				return fmt.Errorf("could not parse owner or repo name from %s", repoArg)
+			}
+		}
+	}
+
+	out := colorableOut(cmd)
+	fmt.Fprintf(out, "Forking %s...\n", utils.Cyan(ghrepo.FullName(toFork)))
+
+	authLogin, err := ctx.AuthLogin()
+	if err != nil {
+		return fmt.Errorf("could not determine current username: %w", err)
+	}
+
+	possibleFork := ghrepo.New(authLogin, toFork.RepoName())
+	exists, err := api.RepoExistsOnGitHub(apiClient, possibleFork)
+	if err != nil {
+		return fmt.Errorf("problem with API request: %w", err)
+	}
+
+	if exists {
+		return fmt.Errorf("%s %s", utils.Cyan(ghrepo.FullName(possibleFork)), utils.Red("already exists!"))
+	}
+
+	forkedRepo, err := api.ForkRepo(apiClient, toFork)
+	if err != nil {
+		return fmt.Errorf("failed to fork: %w", err)
+	}
+
+	fmt.Fprintf(out, "%s %s %s!\n",
+		utils.Cyan(ghrepo.FullName(toFork)),
+		utils.Green("successfully forked to"),
+		utils.Cyan(ghrepo.FullName(forkedRepo)))
+
+	if forceNo {
+		return nil
+	}
+
+	if inParent {
+		remoteDesired := forceYes
+		if !forceYes {
+			err = Confirm("Would you like to add a remote for the new fork?", &remoteDesired)
+			if err != nil {
+				return fmt.Errorf("failed to prompt: %w", err)
+			}
+		}
+		if remoteDesired {
+			_, err := git.AddRemote("fork", forkedRepo.CloneURL, "")
+			if err != nil {
+				return fmt.Errorf("failed to add remote: %w", err)
+			}
+
+			fetchCmd := git.GitCommand("fetch", "fork")
```

Comment:
If a fork was just created, then immediately fetching from it is likely to fail before the fork is "ready"

Reply:
i was worried about that but tested manually a ton without issue. i can add some retry logic though.

Reply:
Two little tricks you can resort to here:
1. If you want to do remote add + fetch, you can do so via a single command: `git remote add -f` instead of via two commands
2. If you want to ensure that the fetch succeeds, even if the fork isn't stricly ready yet, you can intially set the remote URL to be the same as origin remote URL, and then switch it to the fork's URL once the fetch has completed; e.g.
    ```sh
    git remote add -f fork https://github.com/PARENTOWNER/REPO.git
    git remote set-url fork https://github.com/YOURUSER/REPO.git
    ``` 
   This is safe because at the time that the fork is created, its git objects are identical to that of its parent.

Reply:
i spent some time worrying about this too in desktop (https://github.com/desktop/desktop/issues/8949), but i was never able to trigger this possible error case in real world use. after releasing this publicly last month, we have yet to see this error in our issues or exception analytics.

i would strongly recommend against the "using the remote url of the parent" trick. the assumption that the git objects are the same is only valid if you're forking a project for the first time. if you already have a fork for a repo, the github api will return your existing fork without modifying it in any way, which often means its stale (or has extra work) in comparison to the parent.

Reply:
that is comforting to hear as I was also unable to trigger this case locally. I'm very comfortable not worrying about it for now and waiting to see if users report it.